node/network/statement-distribution/src/lib.rs

```diff
@@ -771,18 +845,131 @@ async fn circulate_statement_and_dependents(
 
 fn statement_message(relay_parent: Hash, statement: SignedFullStatement)
 	-> protocol_v1::ValidationProtocol
-{
-	protocol_v1::ValidationProtocol::StatementDistribution(
+{ 
+	let msg = if is_statement_large(&statement) {
+		protocol_v1::StatementDistributionMessage::LargeStatement(
+			StatementMetadata {
+				relay_parent,
+				candidate_hash: statement.payload().candidate_hash(),
+				signed_by: statement.validator_index(),
+				signature: statement.signature().clone(),
+			}
+		)
+	} else {
 		protocol_v1::StatementDistributionMessage::Statement(relay_parent, statement)
-	)
+	};
+
+	protocol_v1::ValidationProtocol::StatementDistribution(msg)
+}
+
+/// Check whether a statement should be treated as large statement.
+fn is_statement_large(statement: &SignedFullStatement) -> bool {
+	match &statement.payload() {
+		Statement::Seconded(committed) => {
+			// Runtime upgrades will always be large and even if not - no harm done.
+			if committed.commitments.new_validation_code.is_some() {
+				return true
+			}
+			// No runtime upgrade, now we need to be more nuanced:
+			let size = committed.commitments.size_estimate();
+
+			// Half max size seems to be a good threshold to start not using notifications:
+			let threshold =
+				PeerSet::Validation.get_info(IsAuthority::Yes)
+					.max_notification_size as usize / 2;
+
+			size >= threshold
+		}
+		Statement::Valid(_) =>
+			false,
+	}
+}
+
+/// Internal trait for estimating encoded size.
+trait SizeEstimate {
+	fn size_estimate(&self) -> usize;
+}
+
+impl<N: SizeEstimate> SizeEstimate for CandidateCommitments<N> {
+	fn size_estimate(&self) -> usize {
```

Comment:
You may find `parity_scale_codec::Encode::encoded_size` useful

Reply:
oh sweet :-)